packages/solarwinds-apm/src/processing/parent.ts

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2023-2024 SolarWinds Worldwide, LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { type Context, type Span, trace } from "@opentelemetry/api"
+import {
+  NoopSpanProcessor,
+  type SpanProcessor,
+} from "@opentelemetry/sdk-trace-base"
+
+import { spanStorage } from "../storage.js"
+
+/** Parent storage where false means no parent */
+const PARENT_STORAGE = spanStorage<Span | false>("solarwinds-apm parent span")
+
+/** Returns true if this span has no parent or its parent is remote */
+export function isRootOrEntry(span: Span): boolean {
+  const parentSpan = PARENT_STORAGE.get(span)
+  return parentSpan === false || parentSpan?.spanContext().isRemote === true
+}
+
+/** Traverses the span hierarchy until the root or entry is found */
```

Comment:
Is this something only made possible by the new span storage?

Reply:
I had a very hacky version of it before that didn't return an actual span but short answer is yes. The goal is to implement transaction names as something like this

```ts
const TXNAME_STORAGE = spanStorage<string>("solarwinds-apm transaction name")

function setTxName(name: string) {
  const span = getRootOrEntry(trace.getActiveSpan())
  TXNAME_STORAGE.set(span, name)
}
```